src/itdelatrisu/opsu/translation/I18n.java

```diff
@@ -0,0 +1,346 @@
+ * (at your option) any later version.
+ *
+ * opsu! is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with opsu!.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package itdelatrisu.opsu.translation;
+
+import itdelatrisu.opsu.Utils;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.IllegalFormatException;
+import java.util.List;
+import java.util.Map;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.regex.Pattern;
+
+import org.newdawn.slick.util.Log;
+
+/**
+ * Provides localisation capabilities to opsu!
+ */
+public class I18n {
+
+	/** A hardcoded instance of the translation framework */
+	private static final I18n FALLBACK;
+
+	/** The list of active languages used to translate opsu! */
+	private static final List<I18n> LOCALES;
+
+	/** An array of String values that represent the identifiers of each locale */
+	public static String[] localeIDs;
+
+	/** The current locale being used to translate opsu! */
+	private static I18n currentLocale;
+
+	/** A mapping of all available translations <code>(key -> value)</code> */
+	private final Map<String, String> keys;
+
+	/**
+	 * Generalized constructor for the translator using input streams
+	 * @param stream A stream that provides data regarding the translation keys
+	 * @param fallbackId A fallback ID to identify this locale if the key <code>language.name</code> is not present.
+	 */
+	private I18n(InputStream stream, String fallbackId) {
+		this.keys = new HashMap<String, String>();
+		if (stream == null)
+			return; // used by the hardcoded fallback translator
+
+		try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
+			String line;
+			while ((line = reader.readLine()) != null) 
+				if (!line.isEmpty() && line.charAt(0) != 35) {
+					String[] values = line.split("=", 2);
+					this.keys.put(values[0], values[1]);
+				}
+			
+		} catch (IOException e) {
+			Log.warn("Could not load all translation keys", e);
+		}
+
+		if (!keys.containsKey("language.name"))
+			keys.put("language.name", fallbackId);
+	}
+
+	/**
+	 * Constructor for language files that can be found outside the JAR
+	 * @param langFile The reference to the external language file
+	 */
+	private I18n(File langFile) throws FileNotFoundException {
+		this(new FileInputStream(langFile), langFile.getName().substring(0, langFile.getName().length() - 5));
```

Comment:
Avoid magic numbers. Make `5` a constant, perhaps something like `LANG_FILE_EXT_LENGTH`, or take the length of a string defined in a constant like `LANG_FILE_EXT`. I would probably opt for the second, especially since it has use later on in `load()`.